ext/drreg/drreg.c

```diff
@@ -1831,115 +1831,198 @@ drreg_is_instr_spill_or_restore(void *drcontext, instr_t *instr, bool *spill OUT
+     * spill regions, ...
+     * The strategy we follow here is to trace the app value of aflags and gpr
+     * as it is moved between spill slots and regs. At times the app value may
+     * be present in the gpr/aflags and spill slot both, but later either one
+     * of the two may be clobbered by a tool write or another spill
+     * respectively. At the end, we restore all live gprs/aflags that do not
+     * have their native app value.
      */
-    uint spilled_to[DR_NUM_GPR_REGS];
-    uint spilled_to_aflags = MAX_SPILLS;
+    uint gpr_spill_slot[DR_NUM_GPR_REGS];
+    uint aflags_spill_slot;
+    reg_id_t aflags_spill_reg;
+    bool gpr_native[DR_NUM_GPR_REGS];
+    bool aflags_native;
+    int last_opcode = 0;
+
     reg_id_t reg;
-    instr_t inst;
-    byte *prev_pc, *pc = info->fragment_info.cache_start_pc;
+    instr_t *inst;
+    byte *pc = info->fragment_info.cache_start_pc;
     uint offs;
     bool spill;
-#ifdef X86
-    bool prev_xax_spill = false;
-    bool aflags_in_xax = false;
-#endif
     uint slot;
     if (pc == NULL)
         return true; /* fault not in cache */
-    for (reg = DR_REG_START_GPR; reg <= DR_REG_STOP_GPR; reg++)
-        spilled_to[GPR_IDX(reg)] = MAX_SPILLS;
+
+    /* At beginning of bb, all gprs and aflags have their native app value. */
+    for (reg = DR_REG_START_GPR; reg <= DR_REG_STOP_GPR; reg++) {
+        gpr_spill_slot[GPR_IDX(reg)] = MAX_SPILLS;
+        gpr_native[GPR_IDX(reg)] = true;
+    }
+    aflags_spill_slot = MAX_SPILLS;
+    aflags_spill_reg = DR_REG_NULL;
+    aflags_native = true;
+
     LOG(drcontext, DR_LOG_ALL, 3,
         "%s: processing fault @" PFX ": decoding from " PFX "\n", __FUNCTION__,
         info->raw_mcontext->pc, pc);
-    instr_init(drcontext, &inst);
-    while (pc < info->raw_mcontext->pc) {
-        instr_reset(drcontext, &inst);
-        prev_pc = pc;
-        pc = decode(drcontext, pc, &inst);
-
-        if (is_our_spill_or_restore(drcontext, &inst, &spill, &reg, &slot, &offs)) {
-            LOG(drcontext, DR_LOG_ALL, 3,
-                "%s @" PFX " found %s to %s offs=0x%x => slot %d\n", __FUNCTION__,
-                prev_pc, spill ? "spill" : "restore", get_register_name(reg), offs, slot);
+    /* XXX: if ilist not available, fall back to some best-effort logic. */
+    ASSERT(info->fragment_info.ilist != NULL, "ilist required for state restoration");
+    for (inst = instrlist_first(info->fragment_info.ilist);
+         inst && pc < info->raw_mcontext->pc; inst = instr_get_next(inst)) {
+        int len = instr_length(drcontext, inst);
+        pc += len;
+        ASSERT(pc <= info->raw_mcontext->pc, "went beyond fault pc while walking ilist");
+        bool app_gpr_restored_now = false;
+        bool app_aflags_restored_now = false;
+        bool app_aflags_written_to_reg_now = false;
+        /* Update book-keeping for gpr/aflags spill slot or aflags spill reg. */
+        if (is_our_spill_or_restore(drcontext, inst, &spill, &reg, &slot, &offs)) {
```

Comment:
Bringing up the discussion about using `is_our_spill_or_restore` or the labels with spill slot metadata added by drreg: I'm re-considering whether we can just use the former for all purposes. The latter seems like overkill in some ways. We discussed the possibility of `is_our_spill_or_restore` being confused if the app uses x5/gs reg in some way that looks like a DR spill/restore. Is that worth adding the extra metadata?

Reply:
I believe it is impossible to be confused on x86, since an app instr that uses %gs will be mangled into a non-far memref.  Unless you're looking at the pre-mangling instrlist?

For aarchxx: it may be possible to be confused if the app uses the stolen reg and the stolen reg mangling swaps which one holds the TLS base and so the mangled version uses the original stolen reg -- maybe confirm?

Reply:
> I believe it is impossible to be confused on x86, since an app instr that uses %gs will be mangled into a non-far memref. Unless you're looking at the pre-mangling instrlist?

`mangle` is set to true for the ilist that is passed to the restore event:
- the multiple `recreate_fragment_ilist` calls at https://github.com/DynamoRIO/dynamorio/blob/4b6d1ca39fc870cd6f20618f65f5d999eb80e734/core/translate.c#L1390; and
-  internally in `recreate_selfmod_ilist` too https://github.com/DynamoRIO/dynamorio/blob/4b6d1ca39fc870cd6f20618f65f5d999eb80e734/core/translate.c#L1146

Checking for AArchXX...

Reply:
>  if the app uses x5/gs reg

Correction for my first comment: AArch64 uses x28 (the stolen reg), not x5 for drreg spill/restores/

I added one test case where the restore state logic doesn't work as intended due to some "faux" spill instr. To fix it, I  simply added a check for whether the instr is a meta/app instr.

Reply:
The next point to consider is: `is_our_spill_or_restore` may return true for non-drreg meta instrs that look like drreg's spills/restores.
- added by non-drreg DR instrumentation like `restore_app_value_to_stolen_reg` during `mangle_stolen_reg`: I see that `is_our_spill_or_restore` returns true for some of those spills, which causes `drreg_event_restore_state` to restore tha app value a second time (already restored at https://github.com/DynamoRIO/dynamorio/blob/4b6d1ca39fc870cd6f20618f65f5d999eb80e734/core/translate.c#L646). 
I think this should be okay though.
- added by client instrumentation that reserved and uses `x28`. Is that even possible?

Reply:
> The next point to consider is: is_our_spill_or_restore may return true for non-drreg meta instrs that look like drreg's spills/restores.

The key thing here is that if drreg is given enough dedicated TLS slots, it will never confuse its spills/restores for another library/tool/DR's.  The only way to be confused is if it doesn't have enough and has to fall back on DR slots.